rpcs3/Emu/Cell/lv2/sys_tty.cpp

```diff
@@ -1,17 +1,87 @@
 #include "stdafx.h"
 #include "sys_tty.h"
 
+#include <deque>
+#include <mutex>
+
 logs::channel sys_tty("sys_tty");
 
 extern fs::file g_tty;
 extern atomic_t<s64> g_tty_size;
+extern std::array<std::deque<std::string>, 16> g_tty_input;
+extern std::mutex g_tty_mutex;
 
 error_code sys_tty_read(s32 ch, vm::ptr<char> buf, u32 len, vm::ptr<u32> preadlen)
 {
-	sys_tty.fatal("sys_tty_read(ch=%d, buf=*0x%x, len=%d, preadlen=*0x%x)", ch, buf, len, preadlen);
+	sys_tty.trace("sys_tty_read(ch=%d, buf=*0x%x, len=%d, preadlen=*0x%x)", ch, buf, len, preadlen);
+
+	if (ch > 15 || !buf)
+	{
+		return CELL_EINVAL;
+	}
+
+	if (ch < SYS_TTYP_USER1)
+	{
+		sys_tty.warning("sys_tty_read called with system channel %d", ch);
+	}
+
+	if (false) // TODO: debug mode check
+	{
+		return CELL_EIO;
```

Comment:
this check should be the first in this function.

Reply:
true